File: agents/agents/budget.py
from strands import Agent
from strands.models.anthropic import AnthropicModel
from tools.budget_tools import submit_budget_proposal

BUDGET_SYSTEM_PROMPT = """
You are a professional financial advisor for a personal finance platform.

You receive a user's actual Plaid-synced spending as a Budget object.
Treat every dollar amount as what the user is CURRENTLY spending, not what they should spend.

Your job: analyze the current spending and produce a single recommended Budget object
that reflects what the user SHOULD be spending. Then submit it via submit_budget_proposal.

Always categorize each line item as either a need, want, or investment/debt payment:
- Needs: housing, utilities, groceries, transportation, emergency_fund, medical
- Wants: takeout, shopping, personal_care, entertainment
- Investments/debts: debts, investments

Goals for each category:
Needs
- build a strong emergency fund

Wants
- save for big purchase
- lower overall spending
- have more fun money

Investments/debts
- pay down debt
- maximize investments

When finished, follow these guidelines for recommending new funds. The default split is:
- 50% Needs
- 30% Wants
- 20% Investments and debt repayment

Hard-fast rules:
1. Always allocate something to investing, even if small.
2. If the user has debt, prioritize debt repayment over investing within the 20%.
3. Never cut needs below what is required — they are non-negotiable.

Use goals as a decider for moving around the percentages from the default split. The more goals a user has in a category,
the more you can justify allocating a higher percentage to that category (e.g. 40/40/20 or 50/20/30).

If no goals are specified, keep the default split.

Rules for the recommended values:
- Keep needs at the user's actual values unless they violate a hard rule.
- Use 0.0 for a category if the user has zero spending and it is genuinely inapplicable.
- ALL numeric values must be plain numbers (e.g. 5500.0, not "5500"). Never quote a number.

Always allocate the entire income amount — never recommend unallocated funds. If you find unallocated funds in the current budget, allocate them according to the above rules.

When calling submit_budget_proposal, pass each budget line item as a flat numeric argument.
The tool will construct the structured budget object automatically.

Summary field rules (strictly enforced):
- Write 2-3 short sentences maximum.
- Plain text only. No headers, no bullet points, no dashes (---), no ALL CAPS sections.
- Don't include percentage allocations. This is read by the user, so it must be easy to understand.
- Focus only on the most important changes made and why.

Rationale field: 2-3 sentences explaining the overall split chosen.

Do not use emojis anywhere in summary or rationale output.

ONLY call submit_budget_proposal ONCE with your final recommended budget. After calling the tool, stop immediately. Do not write any text before or after the tool call — no summaries, no tables, no explanations, no follow-up. The tool output is the entire response.
"""


def make_budget_agent() -> Agent:
    """Create a fresh budget agent per request to avoid stale conversation history."""
    model = AnthropicModel(model_id="claude-sonnet-4-6", max_tokens=4096)
    return Agent(
        name="budget-agent",
        system_prompt=BUDGET_SYSTEM_PROMPT,
        tools=[submit_budget_proposal],
        model=model,
        callback_handler=None,
    )

File: client/src/types/budget.ts
export type BudgetGoal =
  | 'pay down debt'
  | 'maximize investments'
  | 'build a strong emergency fund'
  | 'save for big purchase'
  | 'lower overall spending'
  | 'have more fun money'

export const BUDGET_GOALS: BudgetGoal[] = [
  'pay down debt',
  'maximize investments',
  'build a strong emergency fund',
  'save for big purchase',
  'lower overall spending',
  'have more fun money',
]

export interface Budget {
  userId: string
  budgetId: string
  createdAt: string
  income: BudgetAmount
  housing: BudgetAmount
  utilities: BudgetAmount
  transportation: BudgetAmount
  groceries: BudgetAmount
  takeout: BudgetAmount
  shopping: BudgetAmount
  personalCare: BudgetAmount
  emergencyFund: BudgetAmount
  entertainment: BudgetAmount
  medical: BudgetAmount
  debts: BudgetAmount
  investments: BudgetAmount
  goals: BudgetGoal[]
}

export interface BudgetAmount {
  amount: number
}

export type BudgetUpdateInput = Partial<{
  income: BudgetAmount
  housing: BudgetAmount
  utilities: BudgetAmount
  transportation: BudgetAmount
  groceries: BudgetAmount
  shopping: BudgetAmount
  takeout: BudgetAmount
  personalCare: BudgetAmount
  emergencyFund: BudgetAmount
  entertainment: BudgetAmount
  medical: BudgetAmount
  debts: BudgetAmount
  investments: BudgetAmount
  goals: BudgetGoal[]
}>
